File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/auto/TestFastCycleAuto.java
package org.firstinspires.ftc.teamcode.auto;

import static java.lang.Math.toRadians;

import com.acmerobotics.roadrunner.geometry.Pose2d;
import com.acmerobotics.roadrunner.geometry.Vector2d;
import com.acmerobotics.roadrunner.trajectory.Trajectory;
import com.arcrobotics.ftclib.command.InstantCommand;
import com.arcrobotics.ftclib.command.ParallelDeadlineGroup;
import com.arcrobotics.ftclib.command.SequentialCommandGroup;
import com.arcrobotics.ftclib.command.WaitCommand;
import com.qualcomm.robotcore.eventloop.opmode.Autonomous;

import org.firstinspires.ftc.teamcode.commands.CarouselWheelCommand;
import org.firstinspires.ftc.teamcode.commands.FollowTrajectoryCommand;
import org.firstinspires.ftc.teamcode.commands.FollowTrajectorySequenceCommand;
import org.firstinspires.ftc.teamcode.commands.RelocalizeCommand;
import org.firstinspires.ftc.teamcode.drive.SampleMecanumDrive;
import org.firstinspires.ftc.teamcode.subsystems.CarouselWheel;
import org.firstinspires.ftc.teamcode.subsystems.DistanceSensors;
import org.firstinspires.ftc.teamcode.trajectorysequence.TrajectorySequence;


@Autonomous(name = "Red test cycle")
public class TestFastCycleAuto extends AutoBase {


    private SampleMecanumDrive drive;


    private TrajectorySequence goToHubStart;
    private TrajectorySequence goToWarehouse;
    private TrajectorySequence backToHub;

    private Pose2d startPose = new Pose2d(9.6, -64.0, toRadians(90));

    @Override
    public void initialize() {
        super.initialize();

        drive = new SampleMecanumDrive(hardwareMap);
        drive.setPoseEstimate(startPose);


        telemetry.addLine("Generating trajectories...");
        telemetry.update();

        goToHubStart = drive.trajectorySequenceBuilder(startPose)
                .setReversed(true)
                .splineTo(new Vector2d(-4.5, -40), toRadians(112))
                .waitSeconds(0.3)
                .setReversed(false)
                .build();

        goToWarehouse = drive.trajectorySequenceBuilder(goToHubStart.end())
                .splineToSplineHeading(new Pose2d(9, -62, toRadians(-10)), toRadians(-21))
                .splineToSplineHeading(new Pose2d(19, -63.5, toRadians(0)), toRadians(0))
                .splineTo(new Vector2d(50, -63.5), toRadians(0))
                .build();

        backToHub = drive.trajectorySequenceBuilder(goToWarehouse.end())
                .setReversed(true)
                .splineToSplineHeading(new Pose2d(19, -63.5, toRadians(0)), toRadians(180))
                .splineToSplineHeading(new Pose2d(9, -62, toRadians(-10)), toRadians(159))
                .splineToSplineHeading(new Pose2d(-4.5, -40, toRadians(-75)), toRadians(105))
                .setReversed(false)
                .build();

        telemetry.addLine("Initializing Subsystems...");
        telemetry.update();


        schedule(new SequentialCommandGroup(
                        new InstantCommand(() -> {
                            telemetry.addLine("The program started!");
                            telemetry.update();
                        }),
                        new WaitCommand(1000),
                        new FollowTrajectorySequenceCommand(drive, goToHubStart).andThen(waitFor(300)),
                        new FollowTrajectorySequenceCommand(drive, goToWarehouse).andThen(waitFor(500)),
                        new FollowTrajectorySequenceCommand(drive, backToHub)
                )
        );
    }
}

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/auto/TestFastCycleAuto.kt
package org.firstinspires.ftc.teamcode.auto

import com.acmerobotics.roadrunner.geometry.Pose2d
import com.acmerobotics.roadrunner.geometry.Vector2d
import com.acmerobotics.roadrunner.trajectory.Trajectory
import com.arcrobotics.ftclib.command.CommandOpMode
import com.qualcomm.robotcore.eventloop.opmode.Autonomous
import org.firstinspires.ftc.teamcode.auto.AutoBase
import com.arcrobotics.ftclib.command.InstantCommand
import com.arcrobotics.ftclib.command.SequentialCommandGroup
import com.qualcomm.robotcore.eventloop.opmode.Disabled
import org.firstinspires.ftc.teamcode.commands.FollowTrajectoryCommand
import org.firstinspires.ftc.teamcode.commands.FollowTrajectorySequenceCommand
import org.firstinspires.ftc.teamcode.drive.SampleMecanumDrive
import org.firstinspires.ftc.teamcode.trajectorysequence.TrajectorySequence
import org.firstinspires.ftc.teamcode.util.Extensions.Companion.sendLine
import java.lang.Math.toRadians

@Disabled
@Autonomous(name = "Red test cycle")
class TestFastCycleAuto : AutoBase() {


    private lateinit var goToHubStart: TrajectorySequence
    private lateinit var goToWarehouse: TrajectorySequence
    private lateinit var backToHub: TrajectorySequence


    private lateinit var drive: SampleMecanumDrive

    private val startPose = Pose2d(8.34375, -65.375, toRadians(180.0))


    override fun initialize() {

        super.initialize()


        drive = SampleMecanumDrive(hardwareMap)
        drive.poseEstimate = startPose


        telemetry.sendLine("Generating trajectories...")



        goToHubStart = drive.trajectorySequenceBuilder(startPose)
                .lineToConstantHeading(Vector2d(-10.0, -60.0))
                .waitSeconds(0.3)
                .build()

        goToWarehouse = drive.trajectorySequenceBuilder(goToHubStart.end())
                .setReversed(true)
                .splineToConstantHeading(Vector2d(15.0, -64.0), toRadians(0.0))
                .splineToConstantHeading(Vector2d(50.0, -64.0), toRadians(0.0))
                .build()

        backToHub = drive.trajectorySequenceBuilder(goToWarehouse.end())
                .splineToConstantHeading(Vector2d(15.0, -64.0), toRadians(180.0))
                .splineToConstantHeading(Vector2d(-10.0, -60.0), toRadians(160.0))
                .build()

        telemetry.sendLine("Initializing Subsystems...")


        schedule(SequentialCommandGroup(
                InstantCommand({
                    telemetry.addLine("The program started!")
                    telemetry.update()
                }),
                FollowTrajectorySequenceCommand(drive, goToHubStart),
                FollowTrajectorySequenceCommand(drive, goToWarehouse),
                FollowTrajectorySequenceCommand(drive, backToHub),
                FollowTrajectorySequenceCommand(drive, goToWarehouse),
                FollowTrajectorySequenceCommand(drive, backToHub)
        ))


        telemetry.sendLine("Ready for Start!")
    }
}